patches/server/0932-Add-PlayerPickItemEvent.patch

```diff
@@ -0,0 +1,45 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: RodneyMKay <36546810+RodneyMKay@users.noreply.github.com>
+Date: Wed, 8 Sep 2021 21:34:01 +0200
+Subject: [PATCH] Add PlayerPickItemEvent
+
+
+diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+index 9ad36447e82a3504576f6c10cb297514a2f69289..8d155c16bd53a2db98a19fbb092ab0de73d4b57d 100644
+--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
++++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+@@ -1043,7 +1043,14 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
+             this.disconnect("Invalid hotbar selection (Hacking?)", org.bukkit.event.player.PlayerKickEvent.Cause.ILLEGAL_ACTION); // Paper - kick event cause
+             return;
+         }
+-        this.player.getInventory().pickSlot(packet.getSlot()); // Paper - Diff above if changed
++        Player bukkitPlayer = this.player.getBukkitEntity();
++        int targetSlot = this.player.getInventory().getSuitableHotbarSlot();
++        int sourceSlot = packet.getSlot();
++
++        io.papermc.paper.event.player.PlayerPickItemEvent event = new io.papermc.paper.event.player.PlayerPickItemEvent(bukkitPlayer, targetSlot, sourceSlot);
++        if (!event.callEvent()) return;
++
++        this.player.getInventory().pickSlot(event.getSourceSlot(), event.getTargetSlot());
+         // Paper end
+         this.player.connection.send(new ClientboundContainerSetSlotPacket(-2, 0, this.player.getInventory().selected, this.player.getInventory().getItem(this.player.getInventory().selected)));
+         this.player.connection.send(new ClientboundContainerSetSlotPacket(-2, 0, packet.getSlot(), this.player.getInventory().getItem(packet.getSlot())));
+diff --git a/src/main/java/net/minecraft/world/entity/player/Inventory.java b/src/main/java/net/minecraft/world/entity/player/Inventory.java
+index a1a625a8dacf4d2bbf75ddd90dce1b1be663c919..90e9972819476441576ff7ab45c7a6736a4ec9b4 100644
+--- a/src/main/java/net/minecraft/world/entity/player/Inventory.java
++++ b/src/main/java/net/minecraft/world/entity/player/Inventory.java
+@@ -173,7 +173,13 @@ public class Inventory implements Container, Nameable {
+     }
+ 
+     public void pickSlot(int slot) {
+-        this.selected = this.getSuitableHotbarSlot();
++        // Paper start - Add PlayerPickItemEvent
++        pickSlot(slot, this.getSuitableHotbarSlot());
++    }
++
++    public void pickSlot(int slot, int targetSlot) {
++        this.selected = targetSlot;
```

Comment:
Is this safe to do? change the selected slot?

Reply:
Yes, it seems to work fine.